File: packages/railtracks/src/railtracks/llm/tools/parameters/array_parameter.py
from typing import Any, Dict, Optional, List
from ._base import Parameter, ParameterType

class ArrayParameter(Parameter):
    param_type: ParameterType = ParameterType.ARRAY

    def __init__(
        self,
        name: str,
        items: Parameter,
        description: Optional[str] = None,
        required: bool = True,
        default: Any = None,
        max_items: Optional[int] = None,
        additional_properties: bool = False,
    ):
        super().__init__(name, description, required, default)
        self.items = items
        self.max_items = max_items
        self.additional_properties = additional_properties  # might be relevant if items is object type


    def to_json_schema(self) -> Dict[str, Any]:
        # Base property for items inside the array
        items_schema = self.items.to_json_schema()

        
        schema = {
            "type": "array",
            "items": items_schema,
        }
        if self.description:
            schema["description"] = self.description
            
        if self.max_items is not None:
            schema["maxItems"] = self.max_items

        # Set defaults and enum if present at the array level
        if self.default is not None:
            schema["default"] = self.default

        # Note: enum on arrays is uncommon but if you want to support:
        if self.enum:
            schema["enum"] = self.enum

        return schema

    def __repr__(self) -> str:
        return (
            f"ArrayParameter(name={self.name!r}, items={self.items!r}, "
            f"description={self.description!r}, required={self.required!r}, "
            f"default={self.default!r}, max_items={self.max_items!r}, "
            f"additional_properties={self.additional_properties!r})"
        )

File: dotnet/src/Railtracks.Core/Llm/Tools/ArrayParameter.cs
namespace Railtracks.Llm.Tools;

/// <summary>Parameter representing an array type. Ported from <c>ArrayParameter</c>.</summary>
public sealed class ArrayParameter : Parameter
{
    public ArrayParameter(
        string name,
        Parameter items,
        string? description = null,
        bool required = true,
        object? @default = null,
        int? maxItems = null,
        bool additionalProperties = false)
        : base(name, description, required, @default)
    {
        Items = items;
        MaxItems = maxItems;
        AdditionalProperties = additionalProperties;
        ParamType = ParameterType.Array.Value();
    }

    public Parameter Items { get; }
    public int? MaxItems { get; }
    public bool AdditionalProperties { get; }

    public override Dictionary<string, object?> ToJsonSchema()
    {
        var schema = new Dictionary<string, object?>
        {
            ["type"] = "array",
            ["items"] = Items.ToJsonSchema(),
        };

        if (!string.IsNullOrEmpty(Description))
            schema["description"] = Description;
        if (MaxItems is not null)
            schema["maxItems"] = MaxItems;
        if (Default is not null)
            schema["default"] = Default;
        if (Enum is { Count: > 0 })
            schema["enum"] = Enum;

        return schema;
    }

    public override string ToString() =>
        $"ArrayParameter(name='{Name}', items={Items}, description='{Description}', " +
        $"required={Required}, default={Default}, max_items={MaxItems}, " +
        $"additional_properties={AdditionalProperties})";
}
